File: app/ChallengeEntry.php
<?php

namespace App;

use Illuminate\Database\Eloquent\Model;

class ChallengeEntry extends Model
{
    protected $fillable = ['submission_id', 'challenge_id'];

    public function user() {
        return $this->belongsTo(User::class);
    }

    public function submission() {
        return $this->belongsTo(Submission::class);
    }

    public function challenge() {
        return $this->belongsTo(Challenge::class);
    }

    public function votes() {
        return $this->hasMany(ChallengeVote::class);
    }
}

File: app/models/challenge_entry.rb
# frozen_string_literal: true

class ChallengeEntry < ApplicationRecord
  belongs_to :challenge
  belongs_to :submission
  belongs_to :user

  validates :challenge_id, presence: true
  validates :submission_id, presence: true

  validates :challenge_id, uniqueness: { scope: :submission_id }
end
